OpenCV.js with contrib ximgproc still fails after #28143: EdgeDrawing factory uses unqualified Ptr type

## System information

- OpenCV: `4.x`, `639fd5d07a`
- opencv_contrib: `4.x`, `2a5154a447`
- Emscripten: `6.0.1`
- CMake generator: Ninja
- Target: `opencv.js`
- Build list: `core,imgproc,imgcodecs,video,calib3d,ximgproc,js`

## Detailed description

Building OpenCV.js with `opencv_contrib/modules/ximgproc` enabled fails while
compiling the generated JavaScript binding source.

The generated binding for `cv::ximgproc::EdgeDrawing` uses an unqualified return
type in the factory constructor:

```cpp
.constructor(select_overload<Ptr<EdgeDrawing>()>(&cv::ximgproc::createEdgeDrawing))
```

The class is emitted as `cv::ximgproc::EdgeDrawing`, and the smart pointer
registration also uses the qualified type:

```cpp
emscripten::class_<cv::ximgproc::EdgeDrawing, base<Algorithm>>("ximgproc_EdgeDrawing")
    ...
    .smart_ptr<Ptr<cv::ximgproc::EdgeDrawing>>("Ptr<ximgproc_EdgeDrawing>")
;
```

As a result, the constructor overload should use:

```cpp
Ptr<cv::ximgproc::EdgeDrawing>()
```

## Steps to reproduce

```bash
emcmake cmake \
  -S /path/to/opencv \
  -B /path/to/build \
  -G Ninja \
  -DCMAKE_BUILD_TYPE=Release \
  -DCMAKE_CXX_STANDARD=17 \
  -DOPENCV_EXTRA_MODULES_PATH=/path/to/opencv_contrib/modules \
  -DBUILD_LIST=core,imgproc,imgcodecs,video,calib3d,ximgproc,js \
  -DBUILD_opencv_js=ON \
  -DBUILD_TESTS=OFF \
  -DBUILD_PERF_TESTS=OFF \
  -DBUILD_EXAMPLES=OFF \
  -DBUILD_DOCS=OFF

ninja -C /path/to/build opencv.js
```

## Actual result

```text
modules/js_bindings_generator/gen/bindings.cpp:4455:42: error: unknown type name 'EdgeDrawing'; did you mean 'cv::ximgproc::EdgeDrawing'?
 4455 |         .constructor(select_overload<Ptr<EdgeDrawing>()>(&cv::ximgproc::createEdgeDrawing))
      |                                          ^~~~~~~~~~~
      |                                          cv::ximgproc::EdgeDrawing
```

## Expected result

The OpenCV.js build should complete successfully with `ximgproc` enabled.

The generated constructor binding should use the fully qualified return type:

```cpp
.constructor(select_overload<Ptr<cv::ximgproc::EdgeDrawing>()>(&cv::ximgproc::createEdgeDrawing))
```

## Additional context

This is related to #28130 and #28143.

Although #28130 was closed by #28143, the current `4.x` branch still reproduces
the failure above. The previous fix qualifies factory `Ptr<...>` return types
when the inner type matches the generator class key. In this case the names are:

```text
inner Ptr type:     EdgeDrawing
generator class key: ximgproc_EdgeDrawing
C++ class name:     cv::ximgproc::EdgeDrawing
```

The previous condition does not match this remaining case, so the generated
constructor still uses `Ptr<EdgeDrawing>()`.

--- a/modules/js/generator/embindgen.py
+++ b/modules/js/generator/embindgen.py
@@ -348,7 +348,8 @@ class JSWrapperGenerator(object):
         inner = ret_type[len('Ptr<'):-1].strip()
         if '::' in inner:
             return ret_type
-        if inner == class_info.name:
+        class_basename = class_info.cname.split('::')[-1]
+        if inner == class_info.name or inner == class_basename:
             return 'Ptr<%s>' % class_info.cname
         return ret_type
 

--- a/modules/js/generator/embindgen.py
+++ b/modules/js/generator/embindgen.py
@@ -348,8 +348,7 @@ class JSWrapperGenerator(object):
         inner = ret_type[len('Ptr<'):-1].strip()
         if '::' in inner:
             return ret_type
-        class_basename = class_info.cname.split('::')[-1]
-        if inner == class_info.name or inner == class_basename:
+        if inner == class_info.name or inner == class_info.cname.split('::')[-1]:
             return 'Ptr<%s>' % class_info.cname
         return ret_type
 

--- a/modules/js/generator/embindgen.py
+++ b/modules/js/generator/embindgen.py
@@ -342,6 +342,18 @@ class JSWrapperGenerator(object):
         }
         return tp in string_types
 
+    def _qualify_factory_ptr_return_type(self, ret_type, class_info):
+        if class_info is None:
+            return ret_type
+        if not ret_type.startswith('Ptr<') or not ret_type.endswith('>'):
+            return ret_type
+        inner = ret_type[len('Ptr<'):-1].strip()
+        if '::' in inner:
+            return ret_type
+        if inner == class_info.name:
+            return 'Ptr<%s>' % class_info.cname
+        return ret_type
+
     def _generate_class_properties(self, class_info, class_bindings):
         # Generate bindings for properties
         for prop in class_info.props:

--- a/modules/js/generator/embindgen.py
+++ b/modules/js/generator/embindgen.py
@@ -535,12 +535,8 @@ class JSWrapperGenerator(object):
 
             # Return type
             ret_type = 'void' if variant.rettype.strip() == '' else variant.rettype
-            # FIX: Ensure namespaced smart-pointer return types in factory methods, e.g.:
-            #      Ptr<EdgeDrawing> → Ptr<cv::ximgproc::EdgeDrawing>
             if factory and class_info is not None and ret_type.startswith('Ptr<'):
-                inner = ret_type[len('Ptr<'):-1].strip()
-                if '::' not in inner and inner == class_info.name:
-                    ret_type = 'Ptr<%s>' % class_info.cname
+                ret_type = self._qualify_factory_ptr_return_type(ret_type, class_info)
 
             if ret_type.startswith('Ptr'):  # smart pointer
                 ptr_type = ret_type.replace('Ptr<', '').replace('>', '')

--- a/modules/js/generator/embindgen.py
+++ b/modules/js/generator/embindgen.py
@@ -723,11 +723,8 @@ class JSWrapperGenerator(object):
             ret_type = 'void' if variant.rettype.strip() == '' else variant.rettype
 
             ret_type = ret_type.strip()
-            # Same namespace fix for factory methods: Ptr<EdgeDrawing> -> Ptr<cv::ximgproc::EdgeDrawing>
             if factory and class_info is not None and ret_type.startswith('Ptr<'):
-                inner = ret_type[len('Ptr<'):-1].strip()
-                if '::' not in inner and inner == class_info.name:
-                    ret_type = 'Ptr<%s>' % class_info.cname
+                ret_type = self._qualify_factory_ptr_return_type(ret_type, class_info)
 
             if ret_type.startswith('Ptr'): #smart pointer
                 ptr_type = ret_type.replace('Ptr<', '').replace('>', '')

--- a/modules/js/generator/embindgen.py
+++ b/modules/js/generator/embindgen.py
@@ -343,9 +343,7 @@ class JSWrapperGenerator(object):
         return tp in string_types
 
     def _qualify_factory_ptr_return_type(self, ret_type, class_info):
-        if class_info is None:
-            return ret_type
-        if not ret_type.startswith('Ptr<') or not ret_type.endswith('>'):
+        if class_info is None or not ret_type.startswith('Ptr<') or not ret_type.endswith('>'):
             return ret_type
         inner = ret_type[len('Ptr<'):-1].strip()
         if '::' in inner:

